Restructure form widgets

adammichaelwood opened the pull request:
closes #566 
closes #569 
closes #38 
closes #411 
closes #378 
closes #570 
closes #568 
closes #490 


## What is included?

Near-complete rewrite of `form-widgets.rst`, including new sections for several additional widgets.


## Any problems?

Few things (misspellings, etc.) that I think should have been caught by style checks, but weren't. Will file a few issues against this.

(Before I forget: labelled/labeled, XSLForm)

## Work still to do?

Shrink images a bit more.

lognaturel commented on odk1-src/form-widgets.rst:
```diff
@@ -925,104 +760,304 @@ XLSForm Rows
+.. _single-select-widget:
 
-Single Select Widgets
+Single select widget
```
Wondering whether it might be helpful to have a short introduction to selects with the text you have in the "Including Images as Choices" section below all the select widgets. Then it gives a good sense of the possibilities.

adammichaelwood replied:
that makes sense. I'll try to make that work.

adammichaelwood commented:
> I'd like to see a little bit more narrative to help users figure out non-obvious characteristics of some of the widgets (e.g. the text widget expands vertically) and why certain widgets are useful (e.g. text with numbers appearance is useful for phone numbers).

Great. Will do.

> I'm also finding the "See Also" notes a little disconnected from their section. I'd tend to put them above the screenshots. 

Yeah, I wasn't sure of the best spot. I'll move them up.

> How about scaling them as a percentage of the screen but no bigger than say 300 pixels wide or so? It would be great to get two on the same line where possible.

+1

adammichaelwood commented on odk1-src/form-widgets.rst:
```diff
@@ -13,733 +13,477 @@
-Form Widgets
-***************
 
-This document is a list of available ODK :term:`Collect` :term:`form` :term:`widgets <widget>` (question types), with:
```
Yes. That is coming.

adammichaelwood commented on odk1-src/form-widgets.rst:
```diff
@@ -13,733 +13,477 @@
 
-This section shows examples of all the form widgets types, with no additional options displayed.
+All of the text widgets share the :tc:`text` type,
+and the inputs from them are saved as literal strings.
```
+1

adammichaelwood commented on odk1-src/form-widgets.rst:
```diff
@@ -13,733 +13,477 @@
+appearance
+  *none*
+  
 A simple text input.
```
+1

adammichaelwood commented on odk1-src/form-widgets.rst:
```diff
@@ -13,733 +13,477 @@
 
-Acknowledge
--------------
+Default decimal widget
```
that makes sense

adammichaelwood commented on odk1-src/form-widgets.rst:
```diff
@@ -13,733 +13,477 @@
 
-URL Widget
-~~~~~~~~~~~~
+Range widgets allow the user to select numbers from within a range that is visually represented as a number line. The parameters of the range widget are defined by :tc:`start`, :tc:`end`, and :tc:`step` values defined in the :th:`parameters` column of your XLSForm. The parameter values can be integers or decimals.
```
Great. Will include --- I rely on all-widgets to know what's available...

lognaturel commented:
Sounds great, I look forward to the next iteration!

adammichaelwood commented:
This is ready for review. I still need to run a style check and spelling pass over it myself,
but the things mentioned above have been addressed, and there is some restructure and added wayfinding.

adammichaelwood commented:
note 

- integer widgets capped at 9 digits
- decimal capped at 15

adammichaelwood commented:
note:

 - svg -- inkscape, wikipedia

adammichaelwood commented:
note:
 - shrink images

yanokwa commented:
This is really great, @adammichaelwood. I did find it hard to leave comments inline so my apologies! I've tried to arrange this top down and I'm glad to hop on a call tomorrow if it'd help. 

In the intro, I'd add a hint that you can also use ODK Build to design forms graphically, but it doesn't give you the full range. 

The ordering of the widgets are a bit strange to me. I'd order them as far as frequency of use..
* Text widgets
* Numerical widgets
* Date and time widgets
* Select widgets
* GPS widgets
* Image widgets
* Audio widget
* Video widget
* Barcode widget
* Range widgets
* Note widget
* URL widget
* Printer widget
* Trigger/acknowledge widget
* Grouping multiple questions on the same screen

There is a `multiline` appearance that this doc is missing. It might be good to review http://xlsform.org/ref-table to make sure we have them all.

I think maybe the number string widget should be called the number text widget. We don't use `string` anywhere.

The integer widget has a limit of 9 digits, but the decimal widget has a limit of 14 digits or 15 characters (not digits).

You call it "Numeric appearance attributes" but you don't use attributes as a header elsewhere.

I'd rename GPS widgets to Location. More on this later.

I'd move bearing widget under GPS/Location, and not numeric. I don't think people think of it as a number.

I'd move signature widget up one level. I don't think people think of it as an image.

Under default image widget, you don't mention appearance in the title, but you do so in numeric appearance.  

The barcode app is now called Barcode Scanner, not ZXing.

Audio Record should be Audio Recorder.

Sort the non-Gregorian calendars alphabetically. I don't know how you ordered it but it looks weird to me :)

Single select with search feature can probably drop the feature.

Multiselect should be multi select (I think).

Geopoint doesn't capture GPS always. Sometimes it's WiFi or fused. I'd say Location to be precise. You are missing a period after "Geopoint with user-selected location".

Including images as choices would feel better as the last item in the select list.

In Single Select menu, sometimes you say "Select" other times you "Select widget". I think it should be the latter to be consistent.

In ODK Aggregate, a completed...I'd drop the "In ODK Aggregate". I think it's stored as OK in the form XML.

Trigger acknowledgment widget widget as a repeated word. It might be good to search for "widget widget" to make sure it's not elsewhere.

There is a really big gotcha for grouping multiple question on a screen that should be noted here. See "fields dependent upon an earlier field are not being updated" section on https://opendatakit.org/help/form-design/workarounds/

adammichaelwood commented:
I think i addressed just about everything above.

Couple points:

 - I tested `multiline` and it didn't change the actual appearance *at all*. So I didn't include it. 
 - I'm not entirely happy with the awkwardness of the "questions depending on other questions" note, so if you have some better words for that, please volunteer them
 - i added `meta` as we discussed. Other reference items I've seen put the meta section at the top of the document, but I didn't. I can change that, of course. Also -- i don't know how I feel about wording in there, so feel free to suggest changes. (Always feel free, of course. But, you know -- especially.)
 - I switched "numeric" and "numerical" to "number" in most cases. There's a few where "numeric" seemed to fit better, but you might disagree.

yanokwa commented:
@lognaturel This isn't a blocker for merge, but is it really the case that multiline does nothing on text boxes?

yanokwa commented:
Nitpick, but per our definition, it should be Grouping multiple widgets, not questions on the same screen!

lognaturel commented:
> multiline does nothing on text boxes

That has always been the case. Collect uses the `rows` attribute, Enketo uses the `multiline` appearance. This is less relevant for Collect now because the text control just grows vertically as more text is entered.

adammichaelwood commented:
> per our definition, it should be Grouping multiple widgets, not questions on the same screen!

I thought about that, but then I thought.... maybe the whole screen of questions is one widget...

adammichaelwood commented:
> It will just cause very unpredictable and likely incorrect behavior. 

I would call that a failure condition, no?

yanokwa commented:
Failure implies that there will be something obvious that it has failed. I worry that people will think it's working when they swipe through, but it's failing subtly.

adammichaelwood commented:
do you know if `today` is captured at start, end, completion, or some other time?

yanokwa commented:
Today is captured when the widget is shown.

adammichaelwood commented:
> when the widget is shown.

Except it isn't visible. I'm using "rendered" unless you have a better suggestion...

yanokwa commented on odk1-src/form-widgets.rst:
```diff
@@ -1725,8 +1725,8 @@ and are not visible to the user.
-- :tc:`end` --- The time the survey was completed.
-- :tc:`today` --- The date of survey collection.
+- :tc:`end` --- The last time the survey was saved.
+- :tc:`today` --- The current date, according to the device, when the survey is collected. (The date is determined when the widget is rendered.)
```
Ugh, now that I look at this, the meta widget isn't shown. Let's just drop the stuff in the parenthesis. Better to say nothing than say wrong stuff.
